AssertionError: Unexpected type. Expected 'list' got '<class 'str'>'

### What happened?

```
anton@anton-mbp-m3> pulumi preview                                                                                                                                      ~/tmp/2025-05-21/scala-demo
Previewing update (dev)

View in Browser (Ctrl+O): https://app.pulumi.com/anton-pulumi-corp/scala-demo/dev/previews/8b54c1ef-bbb6-4f77-83a9-284b5d1aaf83

     Type                                         Name                                                     Plan       Info
     pulumi:pulumi:Stack                          scala-demo-dev                                                      1 error
 +   └─ vpcmod:index:Module                       test-vpc                                                 create     
 +      ├─ vpcmod:index:ModuleState               test-vpc-state                                           create     
 +      ├─ vpcmod:tf:aws_route_table_association  module.test-vpc.aws_route_table_association.public[0]    create     
 +      ├─ vpcmod:tf:aws_route_table              module.test-vpc.aws_route_table.private[0]               create     
 +      ├─ vpcmod:tf:aws_internet_gateway         module.test-vpc.aws_internet_gateway.this[0]             create     
 +      ├─ vpcmod:tf:aws_subnet                   module.test-vpc.aws_subnet.database[1]                   create     
 +      ├─ vpcmod:tf:aws_route_table              module.test-vpc.aws_route_table.public[0]                create     
 +      ├─ vpcmod:tf:aws_route                    module.test-vpc.aws_route.public_internet_gateway[0]     create     
 +      ├─ vpcmod:tf:aws_subnet                   module.test-vpc.aws_subnet.public[1]                     create     
 +      ├─ vpcmod:tf:aws_route_table_association  module.test-vpc.aws_route_table_association.public[1]    create     
 +      ├─ vpcmod:tf:aws_default_route_table      module.test-vpc.aws_default_route_table.default[0]       create     
 +      ├─ vpcmod:tf:aws_default_security_group   module.test-vpc.aws_default_security_group.this[0]       create     
 +      ├─ vpcmod:tf:aws_subnet                   module.test-vpc.aws_subnet.public[0]                     create     
 +      ├─ vpcmod:tf:aws_subnet                   module.test-vpc.aws_subnet.private[0]                    create     
 +      ├─ vpcmod:tf:aws_route_table_association  module.test-vpc.aws_route_table_association.private[0]   create     
 +      ├─ vpcmod:tf:aws_route_table              module.test-vpc.aws_route_table.private[1]               create     
 +      ├─ vpcmod:tf:aws_db_subnet_group          module.test-vpc.aws_db_subnet_group.database[0]          create     
 +      ├─ vpcmod:tf:aws_subnet                   module.test-vpc.aws_subnet.private[1]                    create     
 +      ├─ vpcmod:tf:aws_subnet                   module.test-vpc.aws_subnet.database[0]                   create     
 +      ├─ vpcmod:tf:aws_route_table_association  module.test-vpc.aws_route_table_association.private[1]   create     
 +      ├─ vpcmod:tf:aws_route_table_association  module.test-vpc.aws_route_table_association.database[0]  create     
 +      ├─ vpcmod:tf:aws_route_table_association  module.test-vpc.aws_route_table_association.database[1]  create     
 +      ├─ vpcmod:tf:aws_vpc                      module.test-vpc.aws_vpc.this[0]                          create     
 +      └─ vpcmod:tf:aws_default_network_acl      module.test-vpc.aws_default_network_acl.this[0]          create     

Diagnostics:
  pulumi:pulumi:Stack (scala-demo-dev):
    error: Program failed with an unhandled exception:
    Traceback (most recent call last):
      File "/run/current-system/sw/bin/pulumi-language-python-exec", line 232, in <module>
        loop.run_until_complete(coro)
      File "/Users/anton/.local/share/uv/python/cpython-3.12.0-macos-aarch64-none/lib/python3.12/asyncio/base_events.py", line 664, in run_until_complete
        return future.result()
               ^^^^^^^^^^^^^^^
      File "/Users/anton/tmp/2025-05-21/scala-demo/.venv/lib/python3.12/site-packages/pulumi/runtime/stack.py", line 143, in run_in_stack
        await run_pulumi_func(run)
      File "/Users/anton/tmp/2025-05-21/scala-demo/.venv/lib/python3.12/site-packages/pulumi/runtime/stack.py", line 55, in run_pulumi_func
        await wait_for_rpcs()
      File "/Users/anton/tmp/2025-05-21/scala-demo/.venv/lib/python3.12/site-packages/pulumi/runtime/stack.py", line 119, in wait_for_rpcs
        await task
      File "/Users/anton/tmp/2025-05-21/scala-demo/.venv/lib/python3.12/site-packages/pulumi/runtime/resource.py", line 1105, in do_register
        rpc.resolve_outputs(
      File "/Users/anton/tmp/2025-05-21/scala-demo/.venv/lib/python3.12/site-packages/pulumi/runtime/rpc.py", line 1409, in resolve_outputs
        translated_value = translate_output_properties(
                           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
      File "/Users/anton/tmp/2025-05-21/scala-demo/.venv/lib/python3.12/site-packages/pulumi/runtime/rpc.py", line 1285, in translate_output_properties
        element_type = _get_list_element_type(typ)
                       ^^^^^^^^^^^^^^^^^^^^^^^^^^^
      File "/Users/anton/tmp/2025-05-21/scala-demo/.venv/lib/python3.12/site-packages/pulumi/runtime/rpc.py", line 165, in _get_list_element_type
        raise AssertionError(f"Unexpected type. Expected 'list' got '{typ}'")
    AssertionError: Unexpected type. Expected 'list' got '<class 'str'>'

Resources:
    + 24 to create
    1 unchanged

```

### Example

Trying to translate the working TypeScript VPC example to Python:

```python

import pulumi
import pulumi_bucket as bucket
import pulumi_vpcmod as vpcmod
import pulumi_aws as aws
import pulumi_std as std

prefix = pulumi.get_stack()

cidr = "10.0.0.0/16"

azs = aws.get_availability_zones_output(filters=[{
    "name": "opt-in-status",
    "values": ["opt-in-not-required"],
}]).names.apply(lambda names: names[0:2])


def get_cidr_subnet(cidr, netnum):
    return std.cidrsubnet_output(
        input=cidr,
        newbits=8,
        netnum=netnum,
    ).result


my_vpc = vpcmod.Module(
    "test-vpc",
    azs=azs,
    name=f'test-vpcmod',
    cidr=cidr,
    public_subnets=azs.apply(lambda azs: [get_cidr_subnet(cidr, i+1) for (i, az) in enumerate(azs)]),
    private_subnets=azs.apply(lambda azs: [get_cidr_subnet(cidr, i+1+4) for (i, az) in enumerate(azs)]),
    database_subnets=azs.apply(lambda azs: [get_cidr_subnet(cidr, i+1+8) for (i, az) in enumerate(azs)]),
    create_database_subnet_group=True,
)

```

### Output of `pulumi about`

```
CLI          
Version      3.169.0
Go Version   go1.24.3
Go Compiler  gc

Plugins
KIND      NAME              VERSION
resource  aws               6.81.0
language  python            3.169.0
resource  std               2.2.0
resource  terraform-module  0.1.4
resource  terraform-module  0.1.4
resource  terraform-module  0.1.4

Host     
OS       darwin
Version  14.6.1
Arch     arm64

This project is written in python: executable='/Users/anton/tmp/2025-05-21/scala-demo/.venv/bin/python' version='3.12.0'

Current Stack: anton-pulumi-corp/scala-demo/dev

TYPE                                                          URN
pulumi:pulumi:Stack                                           urn:pulumi:dev::scala-demo::pulumi:pulumi:Stack::scala-demo-dev
pulumi:providers:bucket                                       urn:pulumi:dev::scala-demo::pulumi:providers:bucket::default_4_9_0
bucket:index:Module                                           urn:pulumi:dev::scala-demo::bucket:index:Module::my-bucket
pulumi:providers:pulumi                                       urn:pulumi:dev::scala-demo::pulumi:providers:pulumi::default
bucket:index:ModuleState                                      urn:pulumi:dev::scala-demo::bucket:index:Module$bucket:index:ModuleState::my-bucket-state
bucket:tf:aws_s3_bucket_public_access_block                   urn:pulumi:dev::scala-demo::bucket:index:Module$bucket:tf:aws_s3_bucket_public_access_block::module.my-bucket.aws_s3_bucket_public_access_block.this[0]
bucket:tf:aws_s3_bucket_server_side_encryption_configuration  urn:pulumi:dev::scala-demo::bucket:index:Module$bucket:tf:aws_s3_bucket_server_side_encryption_configuration::module.my-bucket.aws_s3_bucket_server_side_encryption_configuration.this[0]
bucket:tf:aws_s3_bucket                                       urn:pulumi:dev::scala-demo::bucket:index:Module$bucket:tf:aws_s3_bucket::module.my-bucket.aws_s3_bucket.this[0]


Found no pending operations associated with dev

Backend        
Name           pulumi.com
URL            https://app.pulumi.com/anton-pulumi-corp
User           anton-pulumi-corp
Organizations  anton-pulumi-corp, moolumi, demo, pulumi
Token type     personal

Dependencies:
NAME           VERSION
pulumi_aws     6.81.0
pulumi_bucket  4.9.0
pulumi_std     2.2.0
pulumi_vpc     5.19.0
pulumi_vpcmod  5.19.0

Pulumi locates its logs in /var/folders/gd/3ncjb1lj5ljgk8xl5ssn_gvc0000gn/T/com.apple.shortcuts.mac-helper// by default

```

### Additional context

N/A

### Contributing

Vote on this issue by adding a 👍 reaction. 
To contribute a fix for this issue, leave a comment (and link to your pull request, if you've opened one already).

Hit this again:

```
Diagnostics:
  pulumi:pulumi:Stack (gcp_module2-dev):
    error: Program failed with an unhandled exception:
    Traceback (most recent call last):
      File "/opt/homebrew/bin/pulumi-language-python-exec", line 232, in <module>
        loop.run_until_complete(coro)
        ~~~~~~~~~~~~~~~~~~~~~~~^^^^^^
      File "/Users/vvm/.pyenv/versions/3.13.0/lib/python3.13/asyncio/base_events.py", line 721, in run_until_complete
        return future.result()
               ~~~~~~~~~~~~~^^
      File "/Users/vvm/code/gcp_module2/venv/lib/python3.13/site-packages/pulumi/runtime/stack.py", line 143, in run_in_stack
        await run_pulumi_func(run)
      File "/Users/vvm/code/gcp_module2/venv/lib/python3.13/site-packages/pulumi/runtime/stack.py", line 55, in run_pulumi_func
        await wait_for_rpcs()
      File "/Users/vvm/code/gcp_module2/venv/lib/python3.13/site-packages/pulumi/runtime/stack.py", line 119, in wait_for_rpcs
        await task
      File "/Users/vvm/code/gcp_module2/venv/lib/python3.13/site-packages/pulumi/runtime/resource.py", line 1105, in do_register
        rpc.resolve_outputs(
        ~~~~~~~~~~~~~~~~~~~^
            res,
            ^^^^
        ...<6 lines>...
            keep_unknowns,
            ^^^^^^^^^^^^^^
        )
        ^
      File "/Users/vvm/code/gcp_module2/venv/lib/python3.13/site-packages/pulumi/runtime/rpc.py", line 1409, in resolve_outputs
        translated_value = translate_output_properties(
            value,
        ...<3 lines>...
            path=_Path(translated_key, resource=f"{res._name}"),
        )
      File "/Users/vvm/code/gcp_module2/venv/lib/python3.13/site-packages/pulumi/runtime/rpc.py", line 1285, in translate_output_properties
        element_type = _get_list_element_type(typ)
      File "/Users/vvm/code/gcp_module2/venv/lib/python3.13/site-packages/pulumi/runtime/rpc.py", line 165, in _get_list_element_type
        raise AssertionError(f"Unexpected type. Expected 'list' got '{typ}'")
    AssertionError: Unexpected type. Expected 'list' got '<class 'str'>'
```

--- a/pkg/modprovider/tfmodules.go
+++ b/pkg/modprovider/tfmodules.go
@@ -208,6 +208,23 @@ func inferExpressionType(expr hcl.Expression) schema.TypeSpec {
 		case "compact":
 			// compact function has return type string[]
 			return arrayType(stringType)
+		case "try":
+			// expressions of format: try(<expr1>, <expr2>, ..., <default>)
+			// we check the last argument to see if it is a string literal or null
+			// if it is, we return a string type
+			if len(functionCall.Args) > 0 {
+				lastArg := functionCall.Args[len(functionCall.Args)-1]
+				switch lastArg := lastArg.(type) {
+				case *hclsyntax.LiteralValueExpr:
+					literalStringOrNull := lastArg.Val.Type().Equals(cty.String) ||
+						lastArg.Val.Type().Equals(cty.DynamicPseudoType)
+					if literalStringOrNull {
+						// try function with a string literal as the last argument
+						// returns a string type
+						return stringType
+					}
+				}
+			}
 		}
 	}
 
@@ -232,7 +249,11 @@ func inferExpressionType(expr hcl.Expression) schema.TypeSpec {
 		return arrayType(stringType)
 	}
 
-	return stringType
+	// default output type is any
+	// language SDKs are very strict when it comes to type schematized
+	// they have to match with the actual type at runtime from terraform
+	// so we use any as a fallback
+	return anyType
 }
 
 // isVariableReference checks if the given expression is a reference to a variable
@@ -495,25 +516,33 @@ func combineInferredModuleSchema(
 
 		// if the output already exists, we need to merge the types
 		existingOutput := inferredSchema.Outputs[name]
-		if output.Ref != "" {
-			existingOutput.Ref = output.Ref
-		}
+
 		if output.Description != "" {
 			existingOutput.Description = output.Description
 		}
 		if output.Secret {
 			existingOutput.Secret = output.Secret
 		}
+
+		if output.Ref != "" {
+			existingOutput.Ref = output.Ref
+			existingOutput.TypeSpec.AdditionalProperties = nil
+			existingOutput.TypeSpec.Items = nil
+			existingOutput.TypeSpec.Type = ""
+		}
 		if output.TypeSpec.Type != "" {
 			existingOutput.TypeSpec.Type = output.TypeSpec.Type
+			existingOutput.TypeSpec.Ref = ""
 		}
 		if output.TypeSpec.Items != nil {
 			existingOutput.TypeSpec.Items = output.TypeSpec.Items
 			existingOutput.TypeSpec.AdditionalProperties = nil
+			existingOutput.TypeSpec.Ref = ""
 		}
 		if output.TypeSpec.AdditionalProperties != nil {
 			existingOutput.TypeSpec.AdditionalProperties = output.TypeSpec.AdditionalProperties
 			existingOutput.TypeSpec.Items = nil
+			existingOutput.TypeSpec.Ref = ""
 		}
 	}
 

--- a/tests/acc_test.go
+++ b/tests/acc_test.go
@@ -1767,6 +1767,21 @@ func Test_LocalModule_RelativePath_OpenTofu(t *testing.T) {
 	t.Logf("%s", previewResult.StdErr+previewResult.StdOut)
 }
 
+// verify that pulumi package get-schema works which also verifies that the
+// generated schema is valid and can be used to generate an SDK
+func TestPulumiPackageGetSchema(t *testing.T) {
+	t.Parallel()
+	localProviderBinPath := ensureCompiledProvider(t)
+	loadedSchema := pulumiGetSchema(t, localProviderBinPath, []string{
+		"terraform-module",
+		"terraform-aws-modules/vpc/aws",
+		"5.18.1",
+		"vpc",
+	})
+
+	assert.NotEmpty(t, loadedSchema, "expected schema to be loaded")
+}
+
 // assertTFStateResourceExists checks if a resource exists in the TF state.
 // packageName should be the name of the package used in `pulumi package add`
 // resourceAddress should be the full TF address of the resource, e.g. "module.test-bucket.aws_s3_bucket.this"
@@ -1925,6 +1940,29 @@ func ensureCompiledProvider(t *testing.T) string {
 	}
 
 	return binPath
+}
+
+func pulumiGetSchema(t *testing.T, localProviderBinPath string, args []string) string {
+	t.Helper()
+
+	localPulumi := filepath.Join(getRoot(t), ".pulumi", "bin", "pulumi")
+
+	if _, err := os.Stat(localPulumi); os.IsNotExist(err) {
+		t.Errorf("This test requires a locally pinned Pulumi CLI; run `make prepare_local_workspace` first")
+	}
+
+	cmd := exec.Command(localPulumi, append([]string{"package", "get-schema"}, args...)...)
+
+	path := os.Getenv("PATH")
+	path = fmt.Sprintf("%s:%s", filepath.Dir(localProviderBinPath), path)
+
+	cmd.Env = append(os.Environ(), fmt.Sprintf("PATH=%s", path))
+	out, err := cmd.CombinedOutput()
+	if err != nil {
+		t.Fatalf("failed to run pulumi package get-schema: %v\n%s", err, out)
+	}
+
+	return string(out)
 }
 
 func pulumiConvert(t *testing.T, localProviderBinPath, sourceDir, targetDir, language string, generateOnly bool) {

--- a/tests/testdata/aws-vpc/go-views/sdks/vpc/vpc/module.go
+++ b/tests/testdata/aws-vpc/go-views/sdks/vpc/vpc/module.go
@@ -41,7 +41,7 @@ type Module struct {
 	// Name of database subnet group
 	Database_subnet_group_name pulumi.StringPtrOutput `pulumi:"database_subnet_group_name"`
 	// A list of all database subnets, containing the full objects.
-	Database_subnet_objects pulumi.StringPtrOutput `pulumi:"database_subnet_objects"`
+	Database_subnet_objects pulumi.MapArrayOutput `pulumi:"database_subnet_objects"`
 	// List of IDs of database subnets
 	Database_subnets pulumi.StringArrayOutput `pulumi:"database_subnets"`
 	// List of cidr_blocks of database subnets
@@ -93,7 +93,7 @@ type Module struct {
 	// Name of elasticache subnet group
 	Elasticache_subnet_group_name pulumi.StringPtrOutput `pulumi:"elasticache_subnet_group_name"`
 	// A list of all elasticache subnets, containing the full objects.
-	Elasticache_subnet_objects pulumi.StringPtrOutput `pulumi:"elasticache_subnet_objects"`
+	Elasticache_subnet_objects pulumi.MapArrayOutput `pulumi:"elasticache_subnet_objects"`
 	// List of IDs of elasticache subnets
 	Elasticache_subnets pulumi.StringArrayOutput `pulumi:"elasticache_subnets"`
 	// List of cidr_blocks of elasticache subnets
@@ -115,7 +115,7 @@ type Module struct {
 	// List of ARNs of intra subnets
 	Intra_subnet_arns pulumi.StringArrayOutput `pulumi:"intra_subnet_arns"`
 	// A list of all intra subnets, containing the full objects.
-	Intra_subnet_objects pulumi.StringPtrOutput `pulumi:"intra_subnet_objects"`
+	Intra_subnet_objects pulumi.MapArrayOutput `pulumi:"intra_subnet_objects"`
 	// List of IDs of intra subnets
 	Intra_subnets pulumi.StringArrayOutput `pulumi:"intra_subnets"`
 	// List of cidr_blocks of intra subnets
@@ -139,7 +139,7 @@ type Module struct {
 	// List of ARNs of outpost subnets
 	Outpost_subnet_arns pulumi.StringArrayOutput `pulumi:"outpost_subnet_arns"`
 	// A list of all outpost subnets, containing the full objects.
-	Outpost_subnet_objects pulumi.StringPtrOutput `pulumi:"outpost_subnet_objects"`
+	Outpost_subnet_objects pulumi.MapArrayOutput `pulumi:"outpost_subnet_objects"`
 	// List of IDs of outpost subnets
 	Outpost_subnets pulumi.StringArrayOutput `pulumi:"outpost_subnets"`
 	// List of cidr_blocks of outpost subnets
@@ -161,7 +161,7 @@ type Module struct {
 	// List of ARNs of private subnets
 	Private_subnet_arns pulumi.StringArrayOutput `pulumi:"private_subnet_arns"`
 	// A list of all private subnets, containing the full objects.
-	Private_subnet_objects pulumi.StringPtrOutput `pulumi:"private_subnet_objects"`
+	Private_subnet_objects pulumi.MapArrayOutput `pulumi:"private_subnet_objects"`
 	// List of IDs of private subnets
 	Private_subnets pulumi.StringArrayOutput `pulumi:"private_subnets"`
 	// List of cidr_blocks of private subnets
@@ -183,7 +183,7 @@ type Module struct {
 	// List of ARNs of public subnets
 	Public_subnet_arns pulumi.StringArrayOutput `pulumi:"public_subnet_arns"`
 	// A list of all public subnets, containing the full objects.
-	Public_subnet_objects pulumi.StringPtrOutput `pulumi:"public_subnet_objects"`
+	Public_subnet_objects pulumi.MapArrayOutput `pulumi:"public_subnet_objects"`
 	// List of IDs of public subnets
 	Public_subnets pulumi.StringArrayOutput `pulumi:"public_subnets"`
 	// List of cidr_blocks of public subnets
@@ -205,7 +205,7 @@ type Module struct {
 	// ID of redshift subnet group
 	Redshift_subnet_group pulumi.StringPtrOutput `pulumi:"redshift_subnet_group"`
 	// A list of all redshift subnets, containing the full objects.
-	Redshift_subnet_objects pulumi.StringPtrOutput `pulumi:"redshift_subnet_objects"`
+	Redshift_subnet_objects pulumi.MapArrayOutput `pulumi:"redshift_subnet_objects"`
 	// List of IDs of redshift subnets
 	Redshift_subnets pulumi.StringArrayOutput `pulumi:"redshift_subnets"`
 	// List of cidr_blocks of redshift subnets
@@ -228,11 +228,11 @@ type Module struct {
 	// Whether or not the VPC has DNS support
 	Vpc_enable_dns_support pulumi.BoolPtrOutput `pulumi:"vpc_enable_dns_support"`
 	// The ARN of the IAM role used when pushing logs to Cloudwatch log group
-	Vpc_flow_log_cloudwatch_iam_role_arn pulumi.StringPtrOutput `pulumi:"vpc_flow_log_cloudwatch_iam_role_arn"`
+	Vpc_flow_log_cloudwatch_iam_role_arn pulumi.AnyOutput `pulumi:"vpc_flow_log_cloudwatch_iam_role_arn"`
 	// The ARN of the IAM role used when pushing logs cross account
 	Vpc_flow_log_deliver_cross_account_role pulumi.StringPtrOutput `pulumi:"vpc_flow_log_deliver_cross_account_role"`
 	// The ARN of the destination for VPC Flow Logs
-	Vpc_flow_log_destination_arn pulumi.StringPtrOutput `pulumi:"vpc_flow_log_destination_arn"`
+	Vpc_flow_log_destination_arn pulumi.AnyOutput `pulumi:"vpc_flow_log_destination_arn"`
 	// The type of the destination for VPC Flow Logs
 	Vpc_flow_log_destination_type pulumi.StringPtrOutput `pulumi:"vpc_flow_log_destination_type"`
 	// The ID of the Flow Log resource
@@ -1338,8 +1338,8 @@ func (o ModuleOutput) Database_subnet_group_name() pulumi.StringPtrOutput {
 }
 
 // A list of all database subnets, containing the full objects.
-func (o ModuleOutput) Database_subnet_objects() pulumi.StringPtrOutput {
-	return o.ApplyT(func(v *Module) pulumi.StringPtrOutput { return v.Database_subnet_objects }).(pulumi.StringPtrOutput)
+func (o ModuleOutput) Database_subnet_objects() pulumi.MapArrayOutput {
+	return o.ApplyT(func(v *Module) pulumi.MapArrayOutput { return v.Database_subnet_objects }).(pulumi.MapArrayOutput)
 }
 
 // List of IDs of database subnets
@@ -1468,8 +1468,8 @@ func (o ModuleOutput) Elasticache_subnet_group_name() pulumi.StringPtrOutput {
 }
 
 // A list of all elasticache subnets, containing the full objects.
-func (o ModuleOutput) Elasticache_subnet_objects() pulumi.StringPtrOutput {
-	return o.ApplyT(func(v *Module) pulumi.StringPtrOutput { return v.Elasticache_subnet_objects }).(pulumi.StringPtrOutput)
+func (o ModuleOutput) Elasticache_subnet_objects() pulumi.MapArrayOutput {
+	return o.ApplyT(func(v *Module) pulumi.MapArrayOutput { return v.Elasticache_subnet_objects }).(pulumi.MapArrayOutput)
 }
 
 // List of IDs of elasticache subnets
@@ -1523,8 +1523,8 @@ func (o ModuleOutput) Intra_subnet_arns() pulumi.StringArrayOutput {
 }
 
 // A list of all intra subnets, containing the full objects.
-func (o ModuleOutput) Intra_subnet_objects() pulumi.StringPtrOutput {
-	return o.ApplyT(func(v *Module) pulumi.StringPtrOutput { return v.Intra_subnet_objects }).(pulumi.StringPtrOutput)
+func (o ModuleOutput) Intra_subnet_objects() pulumi.MapArrayOutput {
+	return o.ApplyT(func(v *Module) pulumi.MapArrayOutput { return v.Intra_subnet_objects }).(pulumi.MapArrayOutput)
 }
 
 // List of IDs of intra subnets
@@ -1583,8 +1583,8 @@ func (o ModuleOutput) Outpost_subnet_arns() pulumi.StringArrayOutput {
 }
 
 // A list of all outpost subnets, containing the full objects.
-func (o ModuleOutput) Outpost_subnet_objects() pulumi.StringPtrOutput {
-	return o.ApplyT(func(v *Module) pulumi.StringPtrOutput { return v.Outpost_subnet_objects }).(pulumi.StringPtrOutput)
+func (o ModuleOutput) Outpost_subnet_objects() pulumi.MapArrayOutput {
+	return o.ApplyT(func(v *Module) pulumi.MapArrayOutput { return v.Outpost_subnet_objects }).(pulumi.MapArrayOutput)
 }
 
 // List of IDs of outpost subnets
@@ -1638,8 +1638,8 @@ func (o ModuleOutput) Private_subnet_arns() pulumi.StringArrayOutput {
 }
 
 // A list of all private subnets, containing the full objects.
-func (o ModuleOutput) Private_subnet_objects() pulumi.StringPtrOutput {
-	return o.ApplyT(func(v *Module) pulumi.StringPtrOutput { return v.Private_subnet_objects }).(pulumi.StringPtrOutput)
+func (o ModuleOutput) Private_subnet_objects() pulumi.MapArrayOutput {
+	return o.ApplyT(func(v *Module) pulumi.MapArrayOutput { return v.Private_subnet_objects }).(pulumi.MapArrayOutput)
 }
 
 // List of IDs of private subnets
@@ -1693,8 +1693,8 @@ func (o ModuleOutput) Public_subnet_arns() pulumi.StringArrayOutput {
 }
 
 // A list of all public subnets, containing the full objects.
-func (o ModuleOutput) Public_subnet_objects() pulumi.StringPtrOutput {
-	return o.ApplyT(func(v *Module) pulumi.StringPtrOutput { return v.Public_subnet_objects }).(pulumi.StringPtrOutput)
+func (o ModuleOutput) Public_subnet_objects() pulumi.MapArrayOutput {
+	return o.ApplyT(func(v *Module) pulumi.MapArrayOutput { return v.Public_subnet_objects }).(pulumi.MapArrayOutput)
 }
 
 // List of IDs of public subnets
@@ -1748,8 +1748,8 @@ func (o ModuleOutput) Redshift_subnet_group() pulumi.StringPtrOutput {
 }
 
 // A list of all redshift subnets, containing the full objects.
-func (o ModuleOutput) Redshift_subnet_objects() pulumi.StringPtrOutput {
-	return o.ApplyT(func(v *Module) pulumi.StringPtrOutput { return v.Redshift_subnet_objects }).(pulumi.StringPtrOutput)
+func (o ModuleOutput) Redshift_subnet_objects() pulumi.MapArrayOutput {
+	return o.ApplyT(func(v *Module) pulumi.MapArrayOutput { return v.Redshift_subnet_objects }).(pulumi.MapArrayOutput)
 }
 
 // List of IDs of redshift subnets
@@ -1807,8 +1807,8 @@ func (o ModuleOutput) Vpc_enable_dns_support() pulumi.BoolPtrOutput {
 }
 
 // The ARN of the IAM role used when pushing logs to Cloudwatch log group
-func (o ModuleOutput) Vpc_flow_log_cloudwatch_iam_role_arn() pulumi.StringPtrOutput {
-	return o.ApplyT(func(v *Module) pulumi.StringPtrOutput { return v.Vpc_flow_log_cloudwatch_iam_role_arn }).(pulumi.StringPtrOutput)
+func (o ModuleOutput) Vpc_flow_log_cloudwatch_iam_role_arn() pulumi.AnyOutput {
+	return o.ApplyT(func(v *Module) pulumi.AnyOutput { return v.Vpc_flow_log_cloudwatch_iam_role_arn }).(pulumi.AnyOutput)
 }
 
 // The ARN of the IAM role used when pushing logs cross account
@@ -1817,8 +1817,8 @@ func (o ModuleOutput) Vpc_flow_log_deliver_cross_account_role() pulumi.StringPtr
 }
 
 // The ARN of the destination for VPC Flow Logs
-func (o ModuleOutput) Vpc_flow_log_destination_arn() pulumi.StringPtrOutput {
-	return o.ApplyT(func(v *Module) pulumi.StringPtrOutput { return v.Vpc_flow_log_destination_arn }).(pulumi.StringPtrOutput)
+func (o ModuleOutput) Vpc_flow_log_destination_arn() pulumi.AnyOutput {
+	return o.ApplyT(func(v *Module) pulumi.AnyOutput { return v.Vpc_flow_log_destination_arn }).(pulumi.AnyOutput)
 }
 
 // The type of the destination for VPC Flow Logs

--- a/tests/testdata/aws-vpc/go-views/sdks/vpc/vpc/provider.go
+++ b/tests/testdata/aws-vpc/go-views/sdks/vpc/vpc/provider.go
@@ -22,6 +22,11 @@ func NewProvider(ctx *pulumi.Context,
 		args = &ProviderArgs{}
 	}
 
+	if args.Executor == nil {
+		if d := internal.GetEnvOrDefault("", nil, "PULUMI_TERRAFORM_MODULE_EXECUTOR"); d != nil {
+			args.Executor = pulumi.StringPtr(d.(string))
+		}
+	}
 	opts = internal.PkgResourceDefaultOpts(opts)
 	ref, err := internal.PkgGetPackageRef(ctx)
 	if err != nil {
@@ -38,12 +43,16 @@ func NewProvider(ctx *pulumi.Context,
 type providerArgs struct {
 	// provider configuration for aws
 	Aws map[string]interface{} `pulumi:"aws"`
+	// Sets the executor used to run the module.
+	Executor *string `pulumi:"executor"`
 }
 
 // The set of arguments for constructing a Provider resource.
 type ProviderArgs struct {
 	// provider configuration for aws
 	Aws pulumi.MapInput
+	// Sets the executor used to run the module.
+	Executor pulumi.StringPtrInput
 }
 
 func (ProviderArgs) ElementType() reflect.Type {

--- a/tests/testdata/aws-vpc/java-views/sdks/vpc/src/main/java/com/pulumi/vpc/Module.java
+++ b/tests/testdata/aws-vpc/java-views/sdks/vpc/src/main/java/com/pulumi/vpc/Module.java
@@ -205,14 +205,14 @@ public class Module extends com.pulumi.resources.CustomResource {
      * A list of all database subnets, containing the full objects.
      * 
      */
-    @Export(name="database_subnet_objects", refs={String.class}, tree="[0]")
-    private Output</* @Nullable */ String> database_subnet_objects;
+    @Export(name="database_subnet_objects", refs={List.class,Map.class,String.class,Object.class}, tree="[0,[1,2,3]]")
+    private Output</* @Nullable */ List<Map<String,Object>>> database_subnet_objects;
 
     /**
      * @return A list of all database subnets, containing the full objects.
      * 
      */
-    public Output<Optional<String>> database_subnet_objects() {
+    public Output<Optional<List<Map<String,Object>>>> database_subnet_objects() {
         return Codegen.optional(this.database_subnet_objects);
     }
     /**
@@ -569,14 +569,14 @@ public class Module extends com.pulumi.resources.CustomResource {
      * A list of all elasticache subnets, containing the full objects.
      * 
      */
-    @Export(name="elasticache_subnet_objects", refs={String.class}, tree="[0]")
-    private Output</* @Nullable */ String> elasticache_subnet_objects;
+    @Export(name="elasticache_subnet_objects", refs={List.class,Map.class,String.class,Object.class}, tree="[0,[1,2,3]]")
+    private Output</* @Nullable */ List<Map<String,Object>>> elasticache_subnet_objects;
 
     /**
      * @return A list of all elasticache subnets, containing the full objects.
      * 
      */
-    public Output<Optional<String>> elasticache_subnet_objects() {
+    public Output<Optional<List<Map<String,Object>>>> elasticache_subnet_objects() {
         return Codegen.optional(this.elasticache_subnet_objects);
     }
     /**
@@ -723,14 +723,14 @@ public class Module extends com.pulumi.resources.CustomResource {
      * A list of all intra subnets, containing the full objects.
      * 
      */
-    @Export(name="intra_subnet_objects", refs={String.class}, tree="[0]")
-    private Output</* @Nullable */ String> intra_subnet_objects;
+    @Export(name="intra_subnet_objects", refs={List.class,Map.class,String.class,Object.class}, tree="[0,[1,2,3]]")
+    private Output</* @Nullable */ List<Map<String,Object>>> intra_subnet_objects;
 
     /**
      * @return A list of all intra subnets, containing the full objects.
      * 
      */
-    public Output<Optional<String>> intra_subnet_objects() {
+    public Output<Optional<List<Map<String,Object>>>> intra_subnet_objects() {
         return Codegen.optional(this.intra_subnet_objects);
     }
     /**
@@ -891,14 +891,14 @@ public class Module extends com.pulumi.resources.CustomResource {
      * A list of all outpost subnets, containing the full objects.
      * 
      */
-    @Export(name="outpost_subnet_objects", refs={String.class}, tree="[0]")
-    private Output</* @Nullable */ String> outpost_subnet_objects;
+    @Export(name="outpost_subnet_objects", refs={List.class,Map.class,String.class,Object.class}, tree="[0,[1,2,3]]")
+    private Output</* @Nullable */ List<Map<String,Object>>> outpost_subnet_objects;
 
     /**
      * @return A list of all outpost subnets, containing the full objects.
      * 
      */
-    public Output<Optional<String>> outpost_subnet_objects() {
+    public Output<Optional<List<Map<String,Object>>>> outpost_subnet_objects() {
         return Codegen.optional(this.outpost_subnet_objects);
     }
     /**
@@ -1045,14 +1045,14 @@ public class Module extends com.pulumi.resources.CustomResource {
      * A list of all private subnets, containing the full objects.
      * 
      */
-    @Export(name="private_subnet_objects", refs={String.class}, tree="[0]")
-    private Output</* @Nullable */ String> private_subnet_objects;
+    @Export(name="private_subnet_objects", refs={List.class,Map.class,String.class,Object.class}, tree="[0,[1,2,3]]")
+    private Output</* @Nullable */ List<Map<String,Object>>> private_subnet_objects;
 
     /**
      * @return A list of all private subnets, containing the full objects.
      * 
      */
-    public Output<Optional<String>> private_subnet_objects() {
+    public Output<Optional<List<Map<String,Object>>>> private_subnet_objects() {
         return Codegen.optional(this.private_subnet_objects);
     }
     /**
@@ -1199,14 +1199,14 @@ public class Module extends com.pulumi.resources.CustomResource {
      * A list of all public subnets, containing the full objects.
      * 
      */
-    @Export(name="public_subnet_objects", refs={String.class}, tree="[0]")
-    private Output</* @Nullable */ String> public_subnet_objects;
+    @Export(name="public_subnet_objects", refs={List.class,Map.class,String.class,Object.class}, tree="[0,[1,2,3]]")
+    private Output</* @Nullable */ List<Map<String,Object>>> public_subnet_objects;
 
     /**
      * @return A list of all public subnets, containing the full objects.
      * 
      */
-    public Output<Optional<String>> public_subnet_objects() {
+    public Output<Optional<List<Map<String,Object>>>> public_subnet_objects() {
         return Codegen.optional(this.public_subnet_objects);
     }
     /**
@@ -1353,14 +1353,14 @@ public class Module extends com.pulumi.resources.CustomResource {
      * A list of all redshift subnets, containing the full objects.
      * 
      */
-    @Export(name="redshift_subnet_objects", refs={String.class}, tree="[0]")
-    private Output</* @Nullable */ String> redshift_subnet_objects;
+    @Export(name="redshift_subnet_objects", refs={List.class,Map.class,String.class,Object.class}, tree="[0,[1,2,3]]")
+    private Output</* @Nullable */ List<Map<String,Object>>> redshift_subnet_objects;
 
     /**
      * @return A list of all redshift subnets, containing the full objects.
      * 
      */
-    public Output<Optional<String>> redshift_subnet_objects() {
+    public Output<Optional<List<Map<String,Object>>>> redshift_subnet_objects() {
         return Codegen.optional(this.redshift_subnet_objects);
     }
     /**
@@ -1513,14 +1513,14 @@ public class Module extends com.pulumi.resources.CustomResource {
      * The ARN of the IAM role used when pushing logs to Cloudwatch log group
      * 
      */
-    @Export(name="vpc_flow_log_cloudwatch_iam_role_arn", refs={String.class}, tree="[0]")
-    private Output</* @Nullable */ String> vpc_flow_log_cloudwatch_iam_role_arn;
+    @Export(name="vpc_flow_log_cloudwatch_iam_role_arn", refs={Object.class}, tree="[0]")
+    private Output</* @Nullable */ Object> vpc_flow_log_cloudwatch_iam_role_arn;
 
     /**
      * @return The ARN of the IAM role used when pushing logs to Cloudwatch log group
      * 
      */
-    public Output<Optional<String>> vpc_flow_log_cloudwatch_iam_role_arn() {
+    public Output<Optional<Object>> vpc_flow_log_cloudwatch_iam_role_arn() {
         return Codegen.optional(this.vpc_flow_log_cloudwatch_iam_role_arn);
     }
     /**
@@ -1541,14 +1541,14 @@ public class Module extends com.pulumi.resources.CustomResource {
      * The ARN of the destination for VPC Flow Logs
      * 
      */
-    @Export(name="vpc_flow_log_destination_arn", refs={String.class}, tree="[0]")
-    private Output</* @Nullable */ String> vpc_flow_log_destination_arn;
+    @Export(name="vpc_flow_log_destination_arn", refs={Object.class}, tree="[0]")
+    private Output</* @Nullable */ Object> vpc_flow_log_destination_arn;
 
     /**
      * @return The ARN of the destination for VPC Flow Logs
      * 
      */
-    public Output<Optional<String>> vpc_flow_log_destination_arn() {
+    public Output<Optional<Object>> vpc_flow_log_destination_arn() {
         return Codegen.optional(this.vpc_flow_log_destination_arn);
     }
     /**

--- a/tests/testdata/aws-vpc/java-views/sdks/vpc/src/main/java/com/pulumi/vpc/ProviderArgs.java
+++ b/tests/testdata/aws-vpc/java-views/sdks/vpc/src/main/java/com/pulumi/vpc/ProviderArgs.java
@@ -5,6 +5,7 @@ package com.pulumi.vpc;
 
 import com.pulumi.core.Output;
 import com.pulumi.core.annotations.Import;
+import com.pulumi.core.internal.Codegen;
 import java.lang.Object;
 import java.lang.String;
 import java.util.Map;
@@ -32,10 +33,26 @@ public final class ProviderArgs extends com.pulumi.resources.ResourceArgs {
         return Optional.ofNullable(this.aws);
     }
 
+    /**
+     * Sets the executor used to run the module.
+     * 
+     */
+    @Import(name="executor")
+    private @Nullable Output<String> executor;
+
+    /**
+     * @return Sets the executor used to run the module.
+     * 
+     */
+    public Optional<Output<String>> executor() {
+        return Optional.ofNullable(this.executor);
+    }
+
     private ProviderArgs() {}
 
     private ProviderArgs(ProviderArgs $) {
         this.aws = $.aws;
+        this.executor = $.executor;
     }
 
     public static Builder builder() {
@@ -77,7 +94,29 @@ public final class ProviderArgs extends com.pulumi.resources.ResourceArgs {
             return aws(Output.of(aws));
         }
 
+        /**
+         * @param executor Sets the executor used to run the module.
+         * 
+         * @return builder
+         * 
+         */
+        public Builder executor(@Nullable Output<String> executor) {
+            $.executor = executor;
+            return this;
+        }
+
+        /**
+         * @param executor Sets the executor used to run the module.
+         * 
+         * @return builder
+         * 
+         */
+        public Builder executor(String executor) {
+            return executor(Output.of(executor));
+        }
+
         public ProviderArgs build() {
+            $.executor = Codegen.stringProp("executor").output().arg($.executor).env("PULUMI_TERRAFORM_MODULE_EXECUTOR").def("").getNullable();
             return $;
         }
     }

--- a/tests/testdata/aws-vpc/node-views/sdks/vpc/module.ts
+++ b/tests/testdata/aws-vpc/node-views/sdks/vpc/module.ts
@@ -88,7 +88,7 @@ export class Module extends pulumi.CustomResource {
     /**
      * A list of all database subnets, containing the full objects.
      */
-    public /*out*/ readonly database_subnet_objects!: pulumi.Output<string | undefined>;
+    public /*out*/ readonly database_subnet_objects!: pulumi.Output<{[key: string]: any}[] | undefined>;
     /**
      * List of IDs of database subnets
      */
@@ -192,7 +192,7 @@ export class Module extends pulumi.CustomResource {
     /**
      * A list of all elasticache subnets, containing the full objects.
      */
-    public /*out*/ readonly elasticache_subnet_objects!: pulumi.Output<string | undefined>;
+    public /*out*/ readonly elasticache_subnet_objects!: pulumi.Output<{[key: string]: any}[] | undefined>;
     /**
      * List of IDs of elasticache subnets
      */
@@ -236,7 +236,7 @@ export class Module extends pulumi.CustomResource {
     /**
      * A list of all intra subnets, containing the full objects.
      */
-    public /*out*/ readonly intra_subnet_objects!: pulumi.Output<string | undefined>;
+    public /*out*/ readonly intra_subnet_objects!: pulumi.Output<{[key: string]: any}[] | undefined>;
     /**
      * List of IDs of intra subnets
      */
@@ -284,7 +284,7 @@ export class Module extends pulumi.CustomResource {
     /**
      * A list of all outpost subnets, containing the full objects.
      */
-    public /*out*/ readonly outpost_subnet_objects!: pulumi.Output<string | undefined>;
+    public /*out*/ readonly outpost_subnet_objects!: pulumi.Output<{[key: string]: any}[] | undefined>;
     /**
      * List of IDs of outpost subnets
      */
@@ -328,7 +328,7 @@ export class Module extends pulumi.CustomResource {
     /**
      * A list of all private subnets, containing the full objects.
      */
-    public /*out*/ readonly private_subnet_objects!: pulumi.Output<string | undefined>;
+    public /*out*/ readonly private_subnet_objects!: pulumi.Output<{[key: string]: any}[] | undefined>;
     /**
      * List of IDs of private subnets
      */
@@ -372,7 +372,7 @@ export class Module extends pulumi.CustomResource {
     /**
      * A list of all public subnets, containing the full objects.
      */
-    public /*out*/ readonly public_subnet_objects!: pulumi.Output<string | undefined>;
+    public /*out*/ readonly public_subnet_objects!: pulumi.Output<{[key: string]: any}[] | undefined>;
     /**
      * List of IDs of public subnets
      */
@@ -416,7 +416,7 @@ export class Module extends pulumi.CustomResource {
     /**
      * A list of all redshift subnets, containing the full objects.
      */
-    public /*out*/ readonly redshift_subnet_objects!: pulumi.Output<string | undefined>;
+    public /*out*/ readonly redshift_subnet_objects!: pulumi.Output<{[key: string]: any}[] | undefined>;
     /**
      * List of IDs of redshift subnets
      */
@@ -461,7 +461,7 @@ export class Module extends pulumi.CustomResource {
     /**
      * The ARN of the IAM role used when pushing logs to Cloudwatch log group
      */
-    public /*out*/ readonly vpc_flow_log_cloudwatch_iam_role_arn!: pulumi.Output<string | undefined>;
+    public /*out*/ readonly vpc_flow_log_cloudwatch_iam_role_arn!: pulumi.Output<any | undefined>;
     /**
      * The ARN of the IAM role used when pushing logs cross account
      */
@@ -469,7 +469,7 @@ export class Module extends pulumi.CustomResource {
     /**
      * The ARN of the destination for VPC Flow Logs
      */
-    public /*out*/ readonly vpc_flow_log_destination_arn!: pulumi.Output<string | undefined>;
+    public /*out*/ readonly vpc_flow_log_destination_arn!: pulumi.Output<any | undefined>;
     /**
      * The type of the destination for VPC Flow Logs
      */

--- a/tests/testdata/aws-vpc/node-views/sdks/vpc/provider.ts
+++ b/tests/testdata/aws-vpc/node-views/sdks/vpc/provider.ts
@@ -32,6 +32,7 @@ export class Provider extends pulumi.ProviderResource {
         opts = opts || {};
         {
             resourceInputs["aws"] = pulumi.output(args ? args.aws : undefined).apply(JSON.stringify);
+            resourceInputs["executor"] = (args ? args.executor : undefined) ?? (utilities.getEnv("PULUMI_TERRAFORM_MODULE_EXECUTOR") || "");
         }
         opts = pulumi.mergeOptions(utilities.resourceOptsDefaults(), opts);
         super(Provider.__pulumiType, name, resourceInputs, opts, false /*dependency*/, utilities.getPackage());
@@ -46,4 +47,8 @@ export interface ProviderArgs {
      * provider configuration for aws
      */
     aws?: pulumi.Input<{[key: string]: any}>;
+    /**
+     * Sets the executor used to run the module.
+     */
+    executor?: pulumi.Input<string>;
 }

--- a/tests/testdata/aws-vpc/node-views/sdks/vpc/utilities.ts
+++ b/tests/testdata/aws-vpc/node-views/sdks/vpc/utilities.ts
@@ -70,6 +70,7 @@ export function lazyLoad(exports: any, props: string[], loadModule: any) {
     }
 }
 
+/** @internal */
 export async function callAsync<T>(
     tok: string,
     props: pulumi.Inputs,

